### PR TITLE
Support podman image trust command

### PR DESCRIPTION
**Pull request by QiWang19**
implemented:
```
$ podman image trust show --json
$ podman image trust show --raw
$ podman image trust show
$ podman image trust set --type <type> --pubkeysfile <filepath> <registryname>
```

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@mtrmac Could you give this a quick review to make sure @QiWang19 is going in the right path. 
@QiWang19 Could you show the output of the commands you have now?

**Comment by rhatdan** [comment]
Missing man pages,
Missing command completions

**Comment by rhatdan** [comment]
@QiWang19 You need to gofmt your code
And fix the validate stuff.

**Comment by baude** [comment]
Needs tests as well.

File: cmd/podman/trust.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	str := t.(string)
+	typeMap := map[string]string{"insecureAcceptAnything": "accept", "signedBy": "signed", "reject": "reject"}
+	if _, exist := typeMap[str]; !exist {
+		return str, errors.Errorf("Invalid trust type %s", str)
+	}
+	return typeMap[str], nil
+}
+
+func sortMapKeySet(m map[string]map[string]interface{}) []string {
```
Comment (giuseppe) [inline cmd/podman/trust.go]: this function doesn't seem to be used anywhere

File: cmd/podman/trust.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	defer mech.Close()
+	if err := mech.SupportsSigning(); err != nil {
+		return errors.Wrap(err, "Signing not supported")
+	}
+
+	showTrustJSON := c.Bool("json")
+	showTrustRaw := c.Bool("raw")
+
+	var policyJSON interface{}
```
[the remark this answers is not in the document]
Comment (QiWang19) [reply cmd/podman/trust.go]: want to use it to convert map[string]map[stirng] to interface{} for format.out()

File: libpod/runtime.go [hunk 3]
```diff
@@ -794,3 +794,8 @@ func (r *Runtime) generateName() (string, error) {
 func (r *Runtime) ImageRuntime() *image.Runtime {
 	return r.imageRuntime
 }
+
+// ImageContext returns the imagecontext for image resolution
+func (r *Runtime) ImageContext() *types.SystemContext {
```
[the remark this answers is not in the document]
Comment (QiWang19) [reply libpod/runtime.go]: fixed

File: cmd/podman/trust.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		return errors.Wrapf(err, "could not read trust policies")
+	}
+
+	mech, err := signature.NewGPGSigningMechanism()
+	if err != nil {
+		return errors.Wrap(err, "Error initializing GPG")
+	}
+	defer mech.Close()
+	if err := mech.SupportsSigning(); err != nil {
+		return errors.Wrap(err, "Signing not supported")
```
[the remark this answers is not in the document]
Comment (QiWang19) [reply cmd/podman/trust.go]: fixed

**Comment by QiWang19** [comment]
I directly read the file as []byte and use struct， can I continue on this?
@mtrmac PTAL

File: cmd/podman/trust.go [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+	}
+	if c.IsSet("registrypath") {
+		systemRegistriesDirPath = c.String("registrypath")
+	} else {
+		systemRegistriesDirPath = registriesDirPath(runtime.SystemContext())
+	}
+	if c.Bool("raw") {
+		_, err := os.Stdout.Write(policyContent)
+		if err != nil {
+			return errors.Errorf("could not read trust policies")
```
Comment (rhatdan) [inline cmd/podman/trust.go]: You have eaten the error.

return errors.Wrap(err, "could not read trust policies")

File: cmd/podman/trust.go [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+		}
+	} else {
```
Comment (rhatdan) [inline cmd/podman/trust.go]: Remove

File: cmd/podman/trust.go [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+		}
+		policyJSON, err := getPolicyJSON(policyContentStruct, systemRegistriesDirPath)
+		if err != nil {
+			return errors.Wrapf(err, "error reading registry config file")
+		}
+		if c.Bool("json") {
+			var outjson interface{}
+			outjson = policyJSON
+			out := formats.JSONStruct{Output: outjson}
+			formats.Writer(out).Out()
```
Comment (rhatdan) [inline cmd/podman/trust.go]: return formats.Writer(out).Out()

File: cmd/podman/trust.go [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+		} else {
```
Comment (rhatdan) [inline cmd/podman/trust.go]: Remove

File: cmd/podman/trust.go [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+					defaultstruct.Sigstore = repoval["sigstore"].(string)
+				}
+				policyoutput = append(policyoutput, defaultstruct)
+			}
+			var output []interface{}
+			for _, ele := range policyoutput {
+				output = append(output, interface{}(ele))
+			}
+			out := formats.StdoutTemplateArray{Output: output, Template: "{{.Repo}}\t{{.Trusttype}}\t{{.GPGid}}\t{{.Sigstore}}"}
+			formats.Writer(out).Out()
```
Comment (rhatdan) [inline cmd/podman/trust.go]: return formats.Writer(out).Out()

File: cmd/podman/trust.go [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+		}
+	}
+	return nil
```
Comment (rhatdan) [inline cmd/podman/trust.go]: remove

File: cmd/podman/trust.go [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+	if len(args) != 1 {
+		return errors.Errorf("default or a registry name must be specified")
+	}
+	valid, err := isValidImageURI(args[0])
+	if err != nil || !valid {
+		return errors.Wrapf(err, "invalid image uri %s", args[0])
+	}
+
+	var pubkeysfile []string
+	if c.IsSet("pubkeysfile") {
```
Comment (rhatdan) [inline cmd/podman/trust.go]: No need for this check.  Only need.
```
pubkeysfile := c.StringSlice("pubkeysfile")
```

File: cmd/podman/trust.go [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+		pubkeysfile = c.StringSlice("pubkeysfile")
+	}
+	trusttype := "signedBy"
```
Comment (rhatdan) [inline cmd/podman/trust.go]: Set the default in the CLI. then always check.

File: cmd/podman/trust.go [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+	SigStore        string `json:"sigstore"`         // For reading, and if SigStoreStaging is not present, for writing.
+	SigStoreStaging string `json:"sigstore-staging"` // For writing only.
+}
+
+var (
+	setTrustFlags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "type, t",
+			Usage: "Trust type, accept values: signedBy(default), accept, reject.",
+		},
```
Comment (rhatdan) [inline cmd/podman/trust.go]: Value: "signedby"

File: cmd/podman/trust.go [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+	}
+	if len(pubkeysfile) == 0 && trusttype == "signedBy" {
+		return errors.Errorf("At least one public key must be defined for type 'signedBy'")
+	}
+
+	var policyPath string
+	if c.IsSet("policypath") {
+		policyPath = c.String("policypath")
+	} else {
+		policyPath = defaultPolicyPath(runtime.SystemContext())
```
Comment (rhatdan) [inline cmd/podman/trust.go]: Set as Default value above.

File: cmd/podman/trust.go [hunk 15]
```diff
@@ -0,0 +1,557 @@
...
+			logrus.Warnf("error get key identity: %s", err)
+			continue
+		}
+		resultsStr := *(*string)(unsafe.Pointer(&results))
+		scanner := bufio.NewScanner(strings.NewReader(resultsStr))
+		var parseduids []string
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.HasPrefix(line, "uid:") || strings.HasPrefix(line, "pub:") {
+				uid := strings.Split(line, ":")[9]
```
Comment (rhatdan) [inline cmd/podman/trust.go]: This looks a little shaky.

File: cmd/podman/trust.go [hunk 16]
```diff
@@ -0,0 +1,557 @@
...
+func createKeyTmpFile(content []byte) (string, error) {
+	tmpfile, err := ioutil.TempFile("/run/", "")
+	if err != nil {
+		return "", err
+	}
+	if _, err := tmpfile.Write(content); err != nil {
+		return "", err
+
+	}
+	if err := tmpfile.Close(); err != nil {
```
Comment (rhatdan) [inline cmd/podman/trust.go]: This should be a defer tmpfile.Close(), right before the tmpfile.Write.

File: cmd/podman/trust.go [hunk 17]
```diff
@@ -0,0 +1,557 @@
...
+		return "", err
+	}
+	return tmpfile.Name(), nil
+}
+
+var typeDescription = map[string]string{"insecureAcceptAnything": "accept", "signedBy": "signed", "reject": "reject"}
+
+func trustTypeDescription(t string) string {
+	ele, exist := typeDescription[t]
```
Comment (rhatdan) [inline cmd/podman/trust.go]: What is an ele?  Bad variable name.

File: cmd/podman/trust.go [hunk 18]
```diff
@@ -0,0 +1,557 @@
...
+	}
+	if err := tmpfile.Close(); err != nil {
+		return "", err
+	}
+	return tmpfile.Name(), nil
+}
+
+var typeDescription = map[string]string{"insecureAcceptAnything": "accept", "signedBy": "signed", "reject": "reject"}
+
+func trustTypeDescription(t string) string {
```
Comment (rhatdan) [inline cmd/podman/trust.go]: NIT: Specify "type string", do not use single letter variables.

File: cmd/podman/trust.go [hunk 19]
```diff
@@ -0,0 +1,557 @@
...
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+// defaultPolicyPath returns a path to the default policy of the system.
+func defaultPolicyPath(sys *types.SystemContext) string {
+	systemDefaultPolicyPath := "/etc/containers/policy.json"
```
Comment (rhatdan) [inline cmd/podman/trust.go]: We should get this from containers/image, not hard code it here.

File: cmd/podman/trust.go [hunk 20]
```diff
@@ -0,0 +1,557 @@
...
+				policyJSON[repo]["sigstore"] = registryNamespace.SigStore
+			}
+		}
+	}
+	return policyJSON, nil
+}
+
+// registriesDirPath returns a path to registries.d
+func registriesDirPath(sys *types.SystemContext) string {
+	systemRegistriesDirPath := "/etc/containers/registries.d"
```
Comment (rhatdan) [inline cmd/podman/trust.go]: Can we get this path from containers/image?

Comment (QiWang19) [reply cmd/podman/trust.go]: can't find an exported variable, so copied [registriesDirPath(sys *types.SystemContext)](https://github.com/containers/image/blob/9f297120deb84f7da63088cc8e32c541f1b0e963/docker/lookaside.go#L77) from containers/image , the same as defaultPolicyPath.
@mtrmac :thinking:  can I get this path from containers/image?

**Comment by mheon** [comment]
Can we move API-relevant code out of `cmd/podman` into `libpod/image`? I think there's value in a reusable API to retrieve this info.

**Comment by rhatdan** [comment]
I agree

**Comment by QiWang19** [comment]
I only moved `func IsValidImageURI(imguri string) (bool, error)`  into `libpod/image/utils.go`

File: cmd/podman/trust.go (hunk 19 above)
Comment (mheon) [reply cmd/podman/trust.go]: Shouldn't we be using the values `libpod/image` set? If we're working with image trust, we should be mirroring how `libpod` will actually see the images

Comment (QiWang19) [reply cmd/podman/trust.go]: how to use that? the value from `image.Runtime.SignaturePolicyPath`?

Comment (mheon) [reply cmd/podman/trust.go]: Yeah, that would be it

**Comment by QiWang19** [comment]
@mheon I added pkg/trust. how about now, still too much code?

File: cmd/podman/trust.go [hunk 21]
```diff
@@ -0,0 +1,379 @@
...
+		return errors.Wrapf(err, "error setting trust policy")
+	}
+	err = ioutil.WriteFile(policyPath, data, 0644)
+	if err != nil {
+		return errors.Wrapf(err, "error setting trust policy")
+	}
+	return nil
+}
+
+func getGPGId(keys []string) string {
```
Comment (mheon) [inline cmd/podman/trust.go]: Maybe move this to pkg/trust?

Comment (QiWang19) [reply cmd/podman/trust.go]: :ok_hand:

File: cmd/podman/trust.go [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+	i := 0
+	for k := range m {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+func getPolicyJSON(policyContentStruct trust.PolicyContent, systemRegistriesDirPath string) (map[string]map[string]interface{}, error) {
```
Comment (mheon) [inline cmd/podman/trust.go]: I think this can also go to pkg/trust

**Comment by mheon** [comment]
@QiWang19 Two more things I think can afford to move, but looks a lot better now - I get a little worried when Github won't display a file by default because it's too long

**Comment by rhatdan** [comment]
@QiWang19 Is this still WIP?

**Comment by QiWang19** [comment]
@rhatdan PTAL

**Comment by rhatdan** [comment]
LGTM other then some nits on the description. Fix those and we can merge.

File: cmd/podman/trust.go [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+	}
+	var policyoutput []policydefault
+	for _, repo := range sortedRepos {
+		repoval := policyJSON[repo]
+		var defaultstruct policydefault
+		defaultstruct.Repo = repo
+		if repoval["type"] != nil {
+			defaultstruct.Trusttype = trustTypeDescription(repoval["type"].(string))
+		}
+		if repoval["keys"] != nil && len(repoval["keys"].([]string)) > 0 {
```
Comment (mtrmac) [inline cmd/podman/trust.go]: Code like this strongly suggests that `GetPolicyJSON` should not be producing output in the `--json` format directly; instead, the `trust` subpackage should return well-typed data in Go structs conceptually independent form the presentation format (maybe much closer to the `policydefault` data type, or just the original `trust.RepoContent`), and this function should have two different formatters, one into `--json`, one into the other format as `[]policyoutput`.

File: cmd/podman/trust.go [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+	_, err = os.Stat(policyPath)
+	if !os.IsNotExist(err) {
+		policyContent, err := ioutil.ReadFile(policyPath)
+		if err != nil {
+			return errors.Wrapf(err, "unable to read %s", policyPath)
+		}
+		if err := json.Unmarshal(policyContent, &policyContentStruct); err != nil {
+			return errors.Errorf("could not read trust policies")
+		}
+	}
```
Comment (mtrmac) [inline cmd/podman/trust.go]: Does this leave `policyContentStruct.Default` empty when the policy does not exist? That is invalid and it would cause all signature users to reject to process any signatures.

(Granted, it’s unlikely that the file will be missing; but if it is, this code should probably force the user to set the default before accepting any other operations.)

File: pkg/trust/trust.go [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+
+	}
+	return tmpfile.Name(), nil
+}
+
+// GetGPGId return GPG identity, either bracketed <email> or ID string
+// comma separated if more than one key
+func GetGPGId(keys []string) string {
+	for _, k := range keys {
+		if _, err := os.Stat(k); err != nil {
```
Comment (mtrmac) [inline pkg/trust/trust.go]: As mentioned elsewhere, in an API it should _never_ be unclear what a string is supposed to mean.

File: pkg/trust/trust.go [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+			decodeKey, err := base64.StdEncoding.DecodeString(k)
+			if err != nil {
+				logrus.Warnf("error decoding key data")
+				continue
+			}
+			tmpfileName, err := CreateTmpFile("/run/", "", decodeKey)
```
Comment (mtrmac) [inline pkg/trust/trust.go]: `/run` is not writable by non-root. Why not let `ioutil.TempFile` use the default? Or, perhaps, pipe the data into `gpg2` without creating a temporary file at all.

File: pkg/trust/trust.go [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+			defer os.Remove(tmpfileName)
+			k = tmpfileName
+		}
+		cmd := exec.Command("gpg2", "--with-colons", k)
+		results, err := cmd.Output()
+		if err != nil {
+			logrus.Warnf("error get key identity: %s", err)
+			continue
+		}
+		resultsStr := *(*string)(unsafe.Pointer(&results))
```
Comment (mtrmac) [inline pkg/trust/trust.go]: **???!!! What is this??!**. Why is this necessary?
